File: Java/src/stack/Stack.java
package stack;

import java.util.Arrays;

public class Stack<T> {
    private Object[] arr;
    private int top;
    private int size;

    public Stack(int sz) {
        arr = new Object[sz];
        top = -1;
        size = sz;
    }

    public boolean isFull() {
        return size == top + 1;
    }

    public boolean isEmpty() {
        return top == -1;
    }

    public void push(T val) {
        if (isFull()) {
            resize();
        }
        top++;
        arr[top] = val;
    }

    public int getTop() {
        return top;
    }

    public void pop() {
        if (isEmpty()) {
            System.out.println("The stack is empty...!");
            return;
        }
        top--;
    }

    public T pick() {
        if (isEmpty()) {
            System.out.println("The stack is empty...!");
            return null;
        }
        return (T )arr[top];
    }

    public void print() {
        if (isEmpty()) {
            System.out.println("There are no elements to print..!");
            return;
        }
        for (int i = top; i >= 0; i--)
            System.out.print(arr[i] + " ");

        System.out.println();
    }

    private void resize() {
        Object[] newArr = Arrays.copyOf(arr, size * 2);
        size = size * 2;
        arr = newArr;
    }

    public void deleteStack() {
        top = -1;
    }

    public void reverse() {
        int j = 0;
        int i = top;
        while (i >= j) {
            T temp = (T) arr[j];
            arr[j] = arr[i];
            arr[i] = temp;
            j++;
            i--;
        }
    }

    public boolean isValidParentheses(String s) {
        int sz = s.length();
        int t = 0;
        for (int i = 1; i < sz; i++) {
            if (s.charAt(i) == '(' || s.charAt(i) == '{' || s.charAt(i) == '[') {
                push((T) Character.valueOf(s.charAt(i)));
                t++;
            } else if (s.charAt(i) == ')' && s.charAt(t) == '(') {
                pop();
            } else if (s.charAt(i) == '{' && s.charAt(t) == '}') {
                pop();
            } else if (s.charAt(i) == ']' && s.charAt(t) == '[') {
                pop();
            }
        }
        return isEmpty();
    }

    public String removeAllAdjacentDuplicatesInString(String s) {
        int sz = s.length();

        for (int i = 0; i < sz; i++) {
            if (isEmpty()) {
                push((T) Character.valueOf(s.charAt(i)));
                continue;
            }
            if (s.charAt(i) == (char) arr[top]) {
                pop();
            } else {
                push((T) Character.valueOf(s.charAt(i)));
            }
        }
        StringBuilder newString = new StringBuilder(top + 1);
        for (int i = 0; i <= top; i++) {
            newString.append(arr[i]);
        }
        return newString.toString();
    }

    public int  asteroidCollision(int[] array, int sz) {
        Stack<Integer> stack = new Stack<>();

        for (int i = 0; i < sz; i++) {
            if (array[i] < 0) {
                if (stack.isEmpty()) {
                    stack.push(array[i]);
                } else {
                    boolean tregar = true;

                    while (!stack.isEmpty()) {
                        if (stack.peek() > 0) {
                            if (Math.abs(array[i]) > stack.peek()) {
                                stack.pop();
                            } else if (Math.abs(array[i]) == stack.peek()) {
                                stack.pop();
                                tregar = false;
                                break;
                            } else if (Math.abs(array[i]) < stack.peek()) {
                                tregar = false;
                                break;
                            }
                        } else {
                            stack.push(array[i]);
                            tregar = false;
                            break;
                        }
                    }
                    if (tregar) {
                        stack.push(array[i]);
                    }
                }
            } else if (array[i] > 0) {
                stack.push(array[i]);
            }
        }

        // Print the result
        System.out.print("Result: ");
        while (!stack.isEmpty()) {
            System.out.print(stack.pop() + " ");
        }
    }

    public void printFront() {
        for (int i = 0; i <= top; i++)
            System.out.print(arr[i] + " ");
        System.out.println();
    }

    public void insertAtBottom(T val, int pos) {
        if (pos == -1) {
            arr[0] = val;
            top++;
            return;
        }
        arr[pos + 1] = arr[pos];
        insertAtBottom(val, pos - 1);
    }

    public int scoreOfParentheses(String s) {
        int sz = s.length();
        int counter = 0;
        int counter2 = 0;
        for (int i = 0; i < sz; i++) {
            if (s.charAt(i) == '(' && s.charAt(i + 1) == ')') {
                counter++;
                i++;
            }

            if (s.charAt(i) == '(' && s.charAt(i + 1) == '(') {
                while (s.charAt(i) != ')' && s.charAt(i + 1) != ')') {
                    counter2++;
                    i += 2;
                }
                break;
            }
        }

        return (counter2 * 2) + counter;
    }
    
    public int getsize(){
        return size;
    }

}

File: DSA/Stack/Stack.py
class Stack:
    def __init__(self, sz):
        self.arr = [0] * sz
        self.top = -1
        self.size = sz

    def is_full(self):
        return self.size == self.top + 1

    def is_empty(self):
        return self.top == -1

    def push(self, val):
        if self.is_full():
            self.resize()
        self.top += 1
        self.arr[self.top] = val

    def get_top(self):
        return self.top

    def resize(self):
        new_arr = [0] * (self.size * 2)
        for i in range(self.size):
            new_arr[i] = self.arr[i]
        self.size = self.size * 2
        self.arr = new_arr

    def pop(self):
        if self.is_empty():
            print("the stack is empty...!")
            return
        self.top -= 1

    def pick(self):
        if self.is_empty():
            print("the stack is empty...!")
            return 0
        return self.arr[self.top]

    def print_stack(self):
        if self.is_empty():
            print("there are no elements to print..!")
            return
        for i in range(self.top, -1, -1):
            print(self.arr[i], end=" ")
        print()

    def delete_stack(self):
        self.top = -1

    def reverse(self):
        j = 0
        i = self.top
        while i >= j:
            temp = self.arr[j]
            self.arr[j] = self.arr[i]
            self.arr[i] = temp
            j += 1
            i -= 1

    def valid_parentheses(self, s):
        sz = len(s)
        t = 0
        for i in range(1, sz):
            if s[i] == '(' or s[i] == '{' or s[i] == '[':
                self.push(s[i])
                t += 1
            elif s[i] == ')' and s[t] == '(':
                self.pop()
            elif s[i] == '{' and s[t] == '}':
                self.pop()
            elif s[i] == ']' and s[t] == '[':
                self.pop()
        return self.is_empty()

    def remove_all_adjacent_duplicates_in_string(self, s):
        sz = len(s)
        for i in range(sz):
            if self.is_empty():
                self.push(s[i])
                continue
            if s[i] == self.arr[self.top]:
                self.pop()
            else:
                self.push(s[i])
        new_string = ''.join(self.arr[:self.top + 1])
        return new_string

    def asteroid_collision(self, array):
        for i in array:
            if i < 0:
                if self.is_empty():
                    self.push(i)
                else:
                    tregar = True
                    while self.top >= 0:
                        if self.arr[self.top] > 0:
                            if abs(i) > self.arr[self.top]:
                                self.pop()
                            elif abs(i) == self.arr[self.top]:
                                self.pop()
                                tregar = False
                                break
                            elif abs(i) < self.arr[self.top]:
                                tregar = False
                                break
                        else:
                            self.push(i)
                            tregar = False
                            break
                    if tregar:
                        self.push(i)
            elif i > 0:
                self.push(i)

    def print_front(self):
        for i in range(self.top + 1):
            print(self.arr[i], end=" ")
        print()

    def insert_at_bottom(self, val, pos):
        if pos == -1:
            self.arr[0] = val
            self.top += 1
            return
        self.arr[pos + 1] = self.arr[pos]
        self.insert_at_bottom(val, pos - 1)

    def score_of_parentheses(self, s):
        counter = 0
        counter2 = 0
        sz = len(s)
        for i in range(sz):
            if s[i] == '(' and s[i + 1] == ')':
                counter += 1
                i += 1
            if s[i] == '(' and s[i + 1] == '(':
                while s[i] != ')' and s[i + 1] != ')':
                    counter2 += 1
                    i += 2
                break
        res = (counter2 * 2) + counter
        return res
